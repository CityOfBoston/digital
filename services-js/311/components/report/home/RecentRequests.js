// @flow
/* eslint react/jsx-no-bind: 0 */

import React from 'react';
import { observable, action, autorun, untracked, computed } from 'mobx';
import { observer } from 'mobx-react';
import { css } from 'glamor';

import type { AppStore } from '../../../data/store';
import type { SearchRequest } from '../../../data/types';
import { HEADER_HEIGHT } from '../../style-constants';

let Velocity;
if (typeof window !== 'undefined') {
  Velocity = require('velocity-animate');
}

const STICKY_SEARCH_STYLE = css({
  position: 'fixed',
  top: HEADER_HEIGHT,
  background: 'white',
});

const CONTAINER_STYLE = css({
  display: 'flex',
  flexDirection: 'column',
});

const REQUEST_STYLE = css({
  display: 'flex',
  cursor: 'pointer',
});

const THUMBNAIL_SYLE = css({
  width: '8rem',
  height: '8rem',
  margin: '0 1rem 0 0',
  flexShrink: 0,
  backgroundSize: 'cover',
});

const REQUEST_INFO_STYLE = css({
  height: '8rem',
  display: 'flex',
  flex: '1 1 0',
  flexDirection: 'column',
  minWidth: 0,
});

const DESCRIPTION_STYLE = css({
  flex: 1,
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  minHeight: 0,
});

const STATUS_COMMON_STYLE = css({
  color: 'white',
  padding: '0.1111rem 0.33333rem',
  marginRight: '0.6666666rem',
});

const STATUS_OPEN_STYLE = css(STATUS_COMMON_STYLE, {
  backgroundColor: '#62A744',
});

const STATUS_CLOSE_STYLE = css(STATUS_COMMON_STYLE, {
  backgroundColor: '#F6A623',
});

export type Props = {
  store: AppStore,
}

@observer
export default class RecentRequests extends React.Component {
  props: Props;

  @observable.ref mainEl: ?HTMLElement = null;
  @observable.ref searchEl: ?HTMLElement = null;
  @observable query: string = '';

  scrollSelectedIntoViewDisposer: ?Function = null;

  componentDidMount() {
    this.scrollSelectedIntoViewDisposer = autorun(this.scrollSelectedIntoView);
  }

  componentWillUnmount() {
    if (this.scrollSelectedIntoViewDisposer) {
      this.scrollSelectedIntoViewDisposer();
    }
  }

  @action.bound
  setMainEl(mainEl: HTMLElement) {
    this.mainEl = mainEl;
  }

  @action.bound
  setSearchEl(searchEl: HTMLElement) {
    this.searchEl = searchEl;
  }

  scrollSelectedIntoView = () => {
    // Keeps us from getting a dependency on props
    const { store } = untracked(() => Object.assign({}, this.props));
    const { selectedRequest, selectedSource } = store.requestSearch;

    if (selectedRequest && this.mainEl && selectedSource !== 'list') {
      const requestEl = this.mainEl.querySelector(`[data-request-id="${selectedRequest.id}"]`);
      const { searchEl } = this;
      if (Velocity && requestEl && searchEl) {
        Velocity(requestEl, 'scroll', { offset: -HEADER_HEIGHT + -searchEl.clientHeight });
      }
    }
  }

  @computed get stickySearch(): boolean {
    const { store } = untracked(() => Object.assign({}, this.props));

    if (!this.mainEl) {
      return false;
    }

    const mainBounds = this.mainEl.getBoundingClientRect();
    return store.ui.scrollY && mainBounds.top <= HEADER_HEIGHT;
  }

  @action.bound
  handleSearchSubmit(ev: SyntheticInputEvent) {
    const { mainEl } = this;
    const { store } = this.props;

    // This actually triggers the search to happen
    store.requestSearch.query = this.query;

    if (Velocity && mainEl) {
      const bounds = mainEl.getBoundingClientRect();
      Velocity(window.document.body, 'scroll', { offset: bounds.top + store.ui.scrollY + -HEADER_HEIGHT });
    }

    ev.preventDefault();
  }

  @action.bound
  handleSearchInput(ev: SyntheticInputEvent) {
    this.query = ev.target.value;
  }

  @action.bound
  handleHoverRequest(request: SearchRequest) {
    const { store: { requestSearch } } = this.props;
    requestSearch.selectedRequest = request;
    requestSearch.selectedSource = 'list';
  }

  @action.bound
  handleUnhoverRequest() {
    const { store: { requestSearch } } = this.props;
    requestSearch.selectedRequest = null;
    requestSearch.selectedSource = null;
  }

  render() {
    const { store: { requestSearch } } = this.props;
    const { results } = requestSearch;

    const containerStyle = {};
    const searchStyle = {};

    if (this.stickySearch && this.searchEl && this.mainEl) {
      containerStyle.paddingTop = this.searchEl.clientHeight;
      searchStyle.width = this.mainEl.clientWidth;
    }

    return (
      <div className={`${CONTAINER_STYLE.toString()}`} ref={this.setMainEl} style={containerStyle}>
        <div className={this.stickySearch && STICKY_SEARCH_STYLE} ref={this.setSearchEl} style={searchStyle}>
          <div className="p-a300">
            <form className="sf sf--y sf--md" acceptCharset="UTF-8" method="get" action="/lookup" onSubmit={this.handleSearchSubmit}>
              <div className="sf-i">
                <input type="text" name="q" placeholder="Search recent casesâ€¦" value={this.query} onInput={this.handleSearchInput} className="sf-i-f" />
                <button className="sf-i-b">Search</button>
              </div>
            </form>
          </div>

          <hr className="hr hr--dash" />
        </div>


        { results.map(this.renderRequest) }
      </div>
    );
  }

  renderRequest = (request: SearchRequest) => {
    const { store: { requestSearch } } = this.props;
    const mediaUrl = request.mediaUrl || '/static/img/311-watermark.svg';

    let statusStyle;
    let statusText;

    if (request.status === 'open') {
      statusStyle = STATUS_OPEN_STYLE;
      statusText = 'Opened';
    } else {
      statusStyle = STATUS_CLOSE_STYLE;
      statusText = 'Closed';
    }

    return (
      <div
        key={request.id}
        data-request-id={request.id}
        className={`p-a300 ${REQUEST_STYLE.toString()}`}
        onMouseEnter={this.handleHoverRequest.bind(null, request)}
        onMouseLeave={this.handleUnhoverRequest}
        style={{
          backgroundColor: (requestSearch.selectedRequest && requestSearch.selectedRequest.id === request.id) ? '#e0e0e0' : 'transparent',
        }}
      >
        <div className={THUMBNAIL_SYLE} style={{ backgroundImage: `url(${mediaUrl})` }} />
        <div className={REQUEST_INFO_STYLE}>
          <h4 className="t--intro t--ellipsis">{request.service.name}</h4>
          <div className={DESCRIPTION_STYLE}>
            { request.description}
          </div>
          <div style={{ paddingTop: 5 }}>
            <span className={`t--upper t--sans ${statusStyle.toString()}`}>{statusText}</span>
            <span className="t--info" style={{ fontSize: 14 }}>{ request.updatedAtRelativeString }</span>
          </div>
        </div>
      </div>
    );
  }
}
