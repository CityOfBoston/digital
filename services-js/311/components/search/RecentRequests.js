// @flow
/* eslint react/jsx-no-bind: 0 */

import React from 'react';
import { observable, action, autorun, untracked, reaction } from 'mobx';
import { observer } from 'mobx-react';
import { css } from 'glamor';

import type { AppStore } from '../../data/store';
import { HEADER_HEIGHT } from '../style-constants';

import LoadingIcons from '../common/LoadingIcons';
import RecentRequestRow from './RecentRequestRow';

let Velocity;
if (typeof window !== 'undefined') {
  Velocity = require('velocity-animate');
}

const CONTAINER_STYLE = css({
  background: 'white',
  // lets us scroll a full height without the footer coming up. There's actually
  // a little slop because of the sticky header.
  minHeight: `calc(100vh - ${HEADER_HEIGHT}px)`,
  position: 'relative',
  width: '40%',
  maxWidth: '35rem',
  display: 'flex',
  flexDirection: 'column',
});

const LOADING_CONTAINER_STYLE = css({
  flex: 1,
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'space-around',
});

const LOADING_WRAPPER_STYLE = css({
  display: 'flex',
  height: 120,
  marginBottom: 200,
});

export type Props = {|
  store: AppStore,
|}

@observer
export default class RecentRequests extends React.Component {
  props: Props;

  @observable.ref mainEl: ?HTMLElement = null;

  scrollSelectedIntoViewDisposer: ?Function;
  updateWidthDisposer: ?Function;

  componentDidMount() {
    const { store: { ui, requestSearch } } = this.props;

    this.scrollSelectedIntoViewDisposer = autorun(this.scrollSelectedIntoView);
    this.updateWidthDisposer = reaction(() => ({
      mainEl: this.mainEl,
      // means we'll get triggered on resizes, which is important
      // for being % width.
      visibleWidth: ui.visibleWidth,
    }),
    ({ mainEl }) => {
      if (mainEl) {
        requestSearch.resultsListWidth = mainEl.clientWidth;
      }
    }, {
      fireImmediately: true,
      name: 'update results list width',
    });
  }

  componentWillUnmount() {
    if (this.scrollSelectedIntoViewDisposer) {
      this.scrollSelectedIntoViewDisposer();
    }

    if (this.updateWidthDisposer) {
      this.updateWidthDisposer();
    }
  }

  handleSearchSubmit = (ev: SyntheticInputEvent) => {
    // nothing to do, because we are auto-searching
    ev.preventDefault();
  }

  @action.bound
  handleSearchInput(ev: SyntheticInputEvent) {
    const { store } = this.props;

    store.requestSearch.query = ev.target.value;
  }

  @action.bound
  setMainEl(mainEl: ?HTMLElement) {
    this.mainEl = mainEl;
  }

  scrollSelectedIntoView = () => {
    // Keeps us from getting a dependency on props
    const { store } = untracked(() => Object.assign({}, this.props));
    const { selectedRequest, selectedSource } = store.requestSearch;

    if (selectedRequest && this.mainEl && selectedSource !== 'list') {
      const requestEl = this.mainEl.querySelector(`[data-request-id="${selectedRequest.id}"]`);
      if (Velocity && requestEl) {
        Velocity(requestEl, 'scroll', { offset: -HEADER_HEIGHT });
      }
    }
  }

  render() {
    const { store: { requestSearch, ui } } = this.props;
    const { results, query, resultsQuery } = requestSearch;

    return (
      <div className={CONTAINER_STYLE} ref={this.setMainEl}>
        <div className="p-a300">
          <form className="sf sf--y sf--md" acceptCharset="UTF-8" method="get" action="/lookup" onSubmit={this.handleSearchSubmit} role="search">
            <div className="sf-i">
              <input type="text" name="q" aria-label="Search field" placeholder="Search by case ID or keywordsâ€¦" value={requestSearch.query} onChange={this.handleSearchInput} className="sf-i-f" />
              <button type="submit" className="sf-i-b">Search</button>
            </div>
          </form>
        </div>

        <h2 className="a11y--h">Search Results</h2>

        { results.length === 0 && query !== resultsQuery && (
          <div className={LOADING_CONTAINER_STYLE}>
            <div className={LOADING_WRAPPER_STYLE}>
              <LoadingIcons initialDelay={0} serverCompatible reduceMotion={ui.reduceMotion} />
            </div>
          </div>
        )}

        { results.map((request) => <RecentRequestRow key={request.id} request={request} requestSearch={requestSearch} ui={ui} />) }
        { results.length === 0 && query === resultsQuery && (
          <div className="p-a300">
            <div className="t--intro">No results found</div>
            <div className="t--info">Try a different search term or move the map to search a different area.</div>
          </div>
        )}
      </div>
    );
  }
}
