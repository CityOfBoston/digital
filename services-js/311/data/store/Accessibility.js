// @flow

import { observable, autorun } from 'mobx';

export default class Accessibility {
  el: ?HTMLElement = null;

  @observable message: string = '';
  @observable interrupt: boolean = false;

  messageListenerDisposer: ?Function = null;

  attach() {
    this.el = document.getElementById('ariaLive');

    this.messageListenerDisposer = autorun(
      () => {
        if (this.el) {
          this.el.innerText = this.message;
          this.el.setAttribute(
            'aria-live',
            this.interrupt ? 'assertive' : 'polite'
          );
        }
      },
      {
        name: 'a11y message listener',
      }
    );
  }

  detatch() {
    this.el = null;

    if (this.messageListenerDisposer) {
      this.messageListenerDisposer();
    }
  }
}
