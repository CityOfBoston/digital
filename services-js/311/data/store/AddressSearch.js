// @flow

import { observable, action, computed, runInAction } from 'mobx';

import type { LoopbackGraphql } from '../dao/loopback-graphql';
import searchAddress from '../dao/search-address';
import reverseGeocode from '../dao/reverse-geocode';
import type { AddressUnit, SearchAddressPlace } from '../types';

// This class co-ordinates location queries between the LocationPopUp pane
// and the LocationMap map. LocationPopUp automatically updates the
// RequestForm (which contains the data that will go into the request) based
// on changes to this instance.
export default class AddressSearch {
  @observable query: string = '';
  @observable searching: boolean = false;
  @observable lastQuery: string = '';
  @observable.ref lastSearchError: ?Error = null;

  @observable.shallow places: ?Array<SearchAddressPlace> = null;
  @observable _currentPlaceIndex: number = -1;
  @observable highlightedPlaceIndex: number = -1;
  @observable _currentUnitIndex: number = 0;
  @observable
  currentReverseGeocodeLocation: ?{ lat: number, lng: number } = null;
  @observable currentReverseGeocodeLocationIsValid: boolean = true;

  @observable mode: 'search' | 'geocode' = 'search';

  // We keep track of whether the constituent seems to be pointing at a location
  // on the map or choosing from the location menu, since we have to choose to
  // send either lat/lng *or* address string to Open311. lat/lng matches the map
  // exactly, but only address string captures the unit.
  @observable intent: 'ADDRESS' | 'LATLNG' = 'ADDRESS';

  // Used so that LocationMap doesn't zoom things under the search box.
  searchPopupWidth: number = 0;

  loopbackGraphql: ?LoopbackGraphql;

  start(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;
  }

  stop() {
    this.loopbackGraphql = null;
  }

  @computed
  get currentPlaceIndex(): number {
    return this._currentPlaceIndex;
  }

  set currentPlaceIndex(currentPlaceIndex: number) {
    this._currentPlaceIndex = currentPlaceIndex;
    this.intent = 'ADDRESS';
  }

  @computed
  get currentUnitIndex(): number {
    return this._currentUnitIndex;
  }

  set currentUnitIndex(currentUnitIndex: number) {
    this._currentUnitIndex = currentUnitIndex;
    this.intent = 'ADDRESS';
  }

  @action
  async search(selectFirst: boolean): Promise<void> {
    if (!this.loopbackGraphql) {
      return;
    }

    this.places = null;
    this.lastQuery = this.query;
    this.lastSearchError = null;
    this.currentReverseGeocodeLocation = null;
    this.mode = 'search';

    if (!this.query) {
      return;
    }

    try {
      this.searching = true;
      const places = await searchAddress(this.loopbackGraphql, this.query);

      runInAction('search - searchAddress success', () => {
        this.searching = false;
        this.places = places;
        this.currentPlaceIndex = places.length === 1 || selectFirst ? 0 : -1;
        this.highlightedPlaceIndex = this.currentPlaceIndex;
        this.currentUnitIndex = 0;
        this.intent = 'ADDRESS';
      });
    } catch (err) {
      runInAction('search - searchAddress error', () => {
        this.searching = false;
        this.lastSearchError = err;
        window._opbeat &&
          !err.server &&
          window._opbeat('captureException', err);
      });
    }
  }

  @computed
  get location(): ?{ lat: number, lng: number } {
    return (
      this.currentReverseGeocodeLocation ||
      (this.currentPlace ? this.currentPlace.location : null)
    );
  }

  set location(location: {| lat: number, lng: number |}) {
    this.query = '';

    if (
      this.location &&
      this.location.lat === location.lat &&
      this.location.lng === location.lng
    ) {
      return;
    }

    if (!this.loopbackGraphql) {
      return;
    }

    this.currentReverseGeocodeLocation = location;
    this.places = null;
    this.mode = 'geocode';
    this.lastQuery = '';
    this.lastSearchError = null;

    this.searching = true;
    reverseGeocode(this.loopbackGraphql, location).then(
      action('reverseGeocode success', place => {
        this.searching = false;

        if (place) {
          // We overwrite the location in the place to stay with the one picked so
          // that the map marker doesnâ€™t move.
          this.places = [
            {
              location,
              address: place.address,
              addressId: place.addressId,
              units: place.units,
            },
          ];
          this.currentPlaceIndex = 0;
          this.highlightedPlaceIndex = 0;
          this.currentUnitIndex = 0;
          this.currentReverseGeocodeLocationIsValid = true;
          this.intent = 'LATLNG';
        } else {
          this.places = [];
          this.currentPlaceIndex = 0;
          this.highlightedPlaceIndex = 0;
          this.currentUnitIndex = 0;
          this.currentReverseGeocodeLocationIsValid = false;
        }
      }),
      action('reverseGeocode error', err => {
        this.searching = false;
        this.lastSearchError = err;
        window._opbeat && window._opbeat('captureException', err);
      })
    );
  }

  @computed
  get currentPlace(): ?SearchAddressPlace {
    return (
      (this.places &&
        this.places.length > this.currentPlaceIndex &&
        this.places[this.currentPlaceIndex]) ||
      null
    );
  }

  @computed
  get currentUnit(): ?AddressUnit {
    return (
      (this.currentPlace &&
        this.currentPlace.units.length > this.currentUnitIndex &&
        this.currentPlace.units[this.currentUnitIndex]) ||
      null
    );
  }

  @computed
  get notFound(): boolean {
    return !!this.places && this.places.length === 0;
  }

  @computed
  get address(): string {
    if (this.currentUnit) {
      return this.currentUnit.address;
    } else if (this.currentPlace) {
      return this.currentPlace.address;
    } else {
      return '';
    }
  }

  @computed
  get addressId(): ?string {
    if (this.currentUnit) {
      return this.currentUnit.addressId;
    } else if (this.currentPlace) {
      return this.currentPlace.addressId;
    } else {
      return null;
    }
  }

  @computed
  get units(): Array<AddressUnit> {
    return this.currentPlace ? this.currentPlace.units : [];
  }
}
