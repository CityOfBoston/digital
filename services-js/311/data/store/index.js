// @flow

import { observable, computed, useStrict } from 'mobx';
import type { IPromiseBasedObservable } from 'mobx-utils';

import type { Service, ServiceSummary, SubmittedRequest } from '../types';
import Question from './Question';

// MobX will enforce that state changes only happen in action blocks.
useStrict(true);

export class ContactInfo {
  @observable required: boolean = false;
  @observable firstName: string = '';
  @observable lastName: string = '';
  @observable email: string = '';
  @observable phone: string = '';

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return !!(this.firstName && this.lastName);
  }
}

export class LocationInfo {
  @observable required: boolean = false;
  @observable address: string = '';
  @observable.shallow location: ?{| lat: number, lng: number |} = null;

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return this.address.length > 0;
  }
}

export class AppStore {
  @observable description: string = '';

  @observable contactInfo: ContactInfo = new ContactInfo();
  @observable locationInfo: LocationInfo = new LocationInfo();
  @observable questions: Question[] = [];

  @observable.shallow serviceSummaries: ServiceSummary[];
  serviceCache: Map<string, Service> = observable.shallowMap({});

  apiKeys: {[service: string]: string} = {};
  isPhone: boolean = false;

  @observable requestSubmission: ?IPromiseBasedObservable<SubmittedRequest> = null;

  @observable.ref _currentService: ?Service = null;
  @observable.ref currentServiceError: ?Object = null;

  @computed get currentService(): ?Service {
    return this._currentService;
  }

  set currentService(service: ?Service) {
    if (service === this._currentService) {
      return;
    }

    this._currentService = service;
    this.currentServiceError = null;
    this.questions = [];

    if (service) {
      try {
        this.contactInfo.required = service.contactRequired;
        this.locationInfo.required = service.locationRequired;
        this.questions = Question.buildQuestions(service.attributes);
      } catch (e) {
        this.currentServiceError = e;
      }
    }
  }

  @computed
  get questionRequirementsMet(): boolean {
    return this.questions.filter((q) => !q.required || !q.visible || q.requirementsMet).length === this.questions.length;
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
