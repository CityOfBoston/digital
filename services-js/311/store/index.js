// @flow
/* global window */
/* eslint no-underscore-dangle: 0 */

import { createStore, compose, applyMiddleware } from 'redux';
import type { Store as ReduxStore } from 'redux';
import { middleware as reduxPackMiddleware } from 'redux-pack';
import thunk from 'redux-thunk';
import inject from 'redux-inject';

import type { RequestAdditions } from '../server/next-handlers';
import { reducer } from './modules';
import type { AppState, AppAction } from './modules';
import makeLoopbackGraphql from './deps/loopback-graphql';
import type { LoopbackGraphql } from './deps/loopback-graphql';

// Re-export these to keep things centralized in the “store” module.
export type State = AppState;
export type Action = AppAction;

export type Store = ReduxStore<State, Action>;

export type Deps = {|
  loopbackGraphql: LoopbackGraphql,
|}

// We implement our own Dispatch type to account for the actions and return
// values provided by our middleware. (The Redux built-in type is from Action
// (which must have a type property) to Action. We need to take Actions that
// don't have properties and return pretty much anything, since that’s what
// thunks can return.
export type Dispatch = (action: Action) => any;

// For the browser, we keep a singleton store that can be re-used across pages.
// A module global is the only mechanism for sharing between Next.js pages.
let browserStore = null;

// Factored out to allow for injecting test dependencies and state while
// maintaining the same middleware.
export function makeStore(deps: ?Deps = null, initialState: ?State = null): Store {
  // Support for Redux Devtools Extension: https://github.com/zalmoxisus/redux-devtools-extension
  const composeEnhancers = (process.browser && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

  const enhancers = composeEnhancers(
    applyMiddleware(
      inject(deps),
      thunk,
      reduxPackMiddleware,
    ),
  );

  if (initialState) {
    return createStore(reducer, initialState, enhancers);
  } else {
    return createStore(reducer, enhancers);
  }
}

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 *
 * req is the Node request (expected to contain injected dependencies for server
 * data fetching) that should be provided when creating the store for the
 * server’s getInitialProps calls. Can be null when server rendering or on the
 * client.
 */
export default function getStore(req: ?RequestAdditions, initialState: ?State = null): Store {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const deps = {
    loopbackGraphql: makeLoopbackGraphql(req),
  };

  const store = makeStore(deps, initialState);

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
