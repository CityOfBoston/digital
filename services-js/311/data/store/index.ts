/* global window */

import { observable, configure as mobxConfigure, action } from 'mobx';
import Router from 'next/router';

import RequestSearch from './RequestSearch';
import AddressSearch from './AddressSearch';
import Ui from './Ui';
import BrowserLocation from './BrowserLocation';
import AllServices from './AllServices';
import {
  FetchGraphql,
  ScreenReaderSupport,
  GaSiteAnalytics,
  RouterListener,
} from '@cityofboston/next-client-common';
import { SiteAnalytics } from '@cityofboston/next-client-common/build/SiteAnalytics';

// MobX will enforce that state changes only happen in action blocks.
mobxConfigure({ enforceActions: true });

export type LanguagePreference = {
  code: string;
  region: string | null;
  quality: number;
};

export class AppStore {
  screenReaderSupport: ScreenReaderSupport = new ScreenReaderSupport();
  requestSearch: RequestSearch = new RequestSearch();
  ui: Ui = new Ui();
  browserLocation: BrowserLocation = new BrowserLocation();
  addressSearch: AddressSearch = new AddressSearch();
  allServices: AllServices = new AllServices();
  siteAnalytics: SiteAnalytics = new GaSiteAnalytics();

  languages: LanguagePreference[] = [];

  @observable
  liveAgentAvailable: boolean = (typeof window !== 'undefined' &&
    window.LIVE_AGENT_AVAILABLE) ||
  false;

  _liveAgentButtonId: string = '';

  get liveAgentButtonId(): string {
    return this._liveAgentButtonId;
  }

  set liveAgentButtonId(liveAgentButtonId: string) {
    this._liveAgentButtonId = liveAgentButtonId;

    if (typeof window !== 'undefined') {
      if (window.liveagent) {
        this.listenForLiveAgentEvents(liveAgentButtonId);
      } else {
        if (!window._laq) {
          window._laq = [];
        }

        window._laq.push(() =>
          this.listenForLiveAgentEvents(liveAgentButtonId)
        );
      }
    }
  }

  listenForLiveAgentEvents(liveAgentButtonId: string) {
    window.liveagent!.addButtonEventHandler(
      liveAgentButtonId,
      this.liveAgentEventHandler
    );
  }

  @action.bound
  liveAgentEventHandler(event: string) {
    switch (event) {
      case 'BUTTON_AVAILABLE':
        this.liveAgentAvailable = true;
        break;
      case 'BUTTON_UNAVAILABLE':
        this.liveAgentAvailable = false;
        break;
      default:
        break;
    }
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(
  fetchGraphql: FetchGraphql | null = null
): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
    store.screenReaderSupport.attach();
    store.ui.attach();
    store.browserLocation.attach(fetchGraphql);

    new RouterListener().attach({
      router: Router,
      siteAnalytics: store.siteAnalytics,
      screenReaderSupport: store.screenReaderSupport,
    });
  }

  return store;
}
