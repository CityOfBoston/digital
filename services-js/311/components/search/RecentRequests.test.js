// @flow

import React from 'react';
import renderer from 'react-test-renderer';

import { AppStore } from '../../data/store';
import type { SearchRequest } from '../../data/types';

import RecentRequests from './RecentRequests';

export const MOCK_REQUEST: SearchRequest = {
  id: '17-000000001',
  service: {
    name: 'Cosmic Intervention',
  },
  description: 'I think that Thanos is here',
  status: 'open',
  address: 'City Hall Plaza, Boston, MA 02131',
  location: {
    lat: 4,
    lng: 5,
  },
  updatedAt: 1490804343,
  updatedAtRelativeString: '4 minutes ago',
  mediaUrl: null,
};

describe('rendering', () => {
  let store;

  beforeEach(() => {
    store = new AppStore();
    store.requestSearch.updateRequestSearchResults({ requests: [MOCK_REQUEST], query: '' });
  });

  test('results loaded', () => {
    const component = renderer.create(
      <RecentRequests store={store} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('with request selected', () => {
    store.requestSearch.selectedRequest = MOCK_REQUEST;

    const component = renderer.create(
      <RecentRequests store={store} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
