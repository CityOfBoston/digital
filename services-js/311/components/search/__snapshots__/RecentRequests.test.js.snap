// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering results loaded 1`] = `
<div
  className={
    Object {
      "data-css-1g83m8j": "",
    }
  }
>
  <div
    className="p-a300 css-17cwx2g"
  >
    <form
      acceptCharset="UTF-8"
      action="/lookup"
      className="sf sf--y sf--md"
      method="get"
      onSubmit={[Function]}
      role="search"
    >
      <div
        className="sf-i"
      >
        <input
          aria-label="Search field"
          className="sf-i-f"
          name="q"
          onChange={[Function]}
          placeholder="Search by case ID or keywords…"
          type="text"
          value=""
        />
        <button
          className="sf-i-b"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <h2
    className="a11y--h"
  >
    Search Results
  </h2>
  <a
    className="p-a300 css-hlpxua"
    data-request-id="17-000000001"
    href="/reports/17-000000001"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "backgroundColor": "transparent",
        "color": "inherit",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-1bo6x5a": "",
        }
      }
      style={
        Object {
          "backgroundImage": "url(/static/img/311-logo-white-on-grey.svg)",
        }
      }
    >
      <div
        className={
          Object {
            "data-css-1bo6x5a": "",
          }
        }
        style={
          Object {
            "backgroundImage": "url()",
          }
        }
      />
    </div>
    <div
      className={
        Object {
          "data-css-1mz2sc3": "",
        }
      }
    >
      <h4
        className="t--intro t--ellipsis css-1fzcvxa"
      >
        Cosmic Intervention
      </h4>
      <div
        className={
          Object {
            "data-css-1p1d70l": "",
          }
        }
      >
        I think that Thanos is here
      </div>
      <div
        style={
          Object {
            "paddingTop": 5,
          }
        }
      >
        <span
          className="t--upper t--sans css-14tkn0p"
        >
          Opened
        </span>
        <span
          className="t--info"
          style={
            Object {
              "fontSize": 14,
            }
          }
        >
          4 minutes ago
        </span>
      </div>
    </div>
  </a>
</div>
`;

exports[`rendering with request selected 1`] = `
<div
  className={
    Object {
      "data-css-1g83m8j": "",
    }
  }
>
  <div
    className="p-a300 css-17cwx2g"
  >
    <form
      acceptCharset="UTF-8"
      action="/lookup"
      className="sf sf--y sf--md"
      method="get"
      onSubmit={[Function]}
      role="search"
    >
      <div
        className="sf-i"
      >
        <input
          aria-label="Search field"
          className="sf-i-f"
          name="q"
          onChange={[Function]}
          placeholder="Search by case ID or keywords…"
          type="text"
          value=""
        />
        <button
          className="sf-i-b"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <h2
    className="a11y--h"
  >
    Search Results
  </h2>
  <a
    className="p-a300 css-hlpxua"
    data-request-id="17-000000001"
    href="/reports/17-000000001"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "color": "inherit",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-1bo6x5a": "",
        }
      }
      style={
        Object {
          "backgroundImage": "url(/static/img/311-logo-grey-on-white.svg)",
        }
      }
    >
      <div
        className={
          Object {
            "data-css-1bo6x5a": "",
          }
        }
        style={
          Object {
            "backgroundImage": "url()",
          }
        }
      />
    </div>
    <div
      className={
        Object {
          "data-css-1mz2sc3": "",
        }
      }
    >
      <h4
        className="t--intro t--ellipsis css-1fzcvxa"
      >
        Cosmic Intervention
      </h4>
      <div
        className={
          Object {
            "data-css-1p1d70l": "",
          }
        }
      >
        I think that Thanos is here
      </div>
      <div
        style={
          Object {
            "paddingTop": 5,
          }
        }
      >
        <span
          className="t--upper t--sans css-14tkn0p"
        >
          Opened
        </span>
        <span
          className="t--info"
          style={
            Object {
              "fontSize": 14,
            }
          }
        >
          4 minutes ago
        </span>
      </div>
    </div>
  </a>
</div>
`;
